plugins/fundchannel.c

```diff
@@ -376,17 +376,29 @@ static struct command_result *json_fundchannel(struct command *cmd,
 {
 	struct funding_req *fr = tal(cmd, struct funding_req);
 	struct json_out *ret;
+	const char *satoshi_str;
+	const char *amount_str;
 
 	if (!param(cmd, buf, params,
 		   p_req("id", param_node_id, &fr->id),
-		   p_req("satoshi", param_string, &fr->funding_str),
+		   p_opt("amount", param_string, &amount_str),
+		   p_opt("satoshi", param_string, &satoshi_str),
 		   p_opt("feerate", param_string, &fr->feerate_str),
 		   p_opt_def("announce", param_bool, &fr->announce_channel, true),
 		   p_opt_def("minconf", param_number, &fr->minconf, 1),
 		   p_opt("utxos", param_string, &fr->utxo_str),
 		   NULL))
 		return command_param_failed();
 
+	if (!amount_str) {
+		if (deprecated_apis || !satoshi_str)
```

Comment:
I think this logic is wrong?  If amount is not specified, *and* we support deprecated APIs, we fail?  Maybe it should be `!deprecated_apis`?

Reply:
Too embarrassed, I confused the meaning of `deprecated_apis`...

Reply:
I also find this is not compatible with the case if `params->type == JSMN_ARRAY`.

Reply:
I believe a technique we used before would be to use both as `p_opt` if `params->type` is an object and to `p_req` if an array instead.

Reply:
I can confirm `p_req(name, cbx, arg)` is used to `Add a required parameter`, and `p_opt(name, cbx, arg)` is used to `Add an optional parameter.  *arg is set to NULL if it isn't found.` (in `common/param.h`). The "req" in `p_req` represents "require"? And "opt" represents "optional"? These make sense. :-)


> I also find this is not compatible with the case if params->type == JSMN_ARRAY.

This is because I find here, also for `close`:
https://github.com/ElementsProject/lightning/blob/9578fb4035d14690bd2ae8d9788cbb05486dbeaa/lightningd/peer_control.c#L1240-L1313

This `else if` block shows if `params->type == JSMN_ARRAY`, `param` will prase the parameters by position.

So for my original code:
```

	if (!param(cmd, buf, params,
		   p_req("id", param_node_id, &fr->id),
		   p_opt("amount", param_string, &amount_str),
                   p_opt("satoshi", param_string, &satoshi_str),
		   p_opt("feerate", param_string, &fr->feerate_str),
		   p_opt_def("announce", param_bool, &fr->announce_channel, true),
		   p_opt_def("minconf", param_number, &fr->minconf, 1),
		   p_opt("utxos", param_string, &fr->utxo_str),
		   NULL))
```

Parsing both `amount` and `satoshi` by position at the same time will lead to the failure.
Now I fix it.